### PR TITLE
[home_task_4]: Home Task 4

### PR DESCRIPTION
**Pull request by AlexanderBugara**
1) a class User that has fields name and age.
2) a class Card, which has a set of methods to check the checksum and show an appropriate error.
Before a user is allowed to check the card checksum, they should introduce themselves and input the age. Users under 13 y.o. should not be able to check the cards. Users above 13 are allowed to input the credit card number and to get the results of checksum validation. 
The program should log all the actions (user names, ages, and the results of their checks) into a log file.

### CODE REVIEW

File: validate.rb [hunk 1]
```diff
@@ -1,54 +1,170 @@
...
+require 'logger'
+require 'tmpdir'
+
 NINE = 9
 TWO = 2
 TEN = 10
 CC_LENGTH = 16
 CC_LENGTH_WITH_SEPARATORS = 19
 CC_INPUT_ERROR_MESSAGE = 'Your cc input has an error'
+MINIMUM_AGE = 13
```
Comment (denchdev) [inline validate.rb]: Move it to User class

File: validate.rb [hunk 2]
```diff
@@ -1,54 +1,170 @@
...
-    end
-    memo + temp_digit
+# To make logs
+class UserWrapper
+  include UserAbstract
+  attr_reader :user, :logger
+
+  def initialize(name, age, card)
+    @user = User.new(name, age, card)
+    tmp_logger = Logger.new File.open('/tmp/ruby_user.log', 'a')
```
Comment (denchdev) [inline validate.rb]: Move it to Logger class to separate file

File: validate.rb [hunk 3]
```diff
@@ -1,54 +1,170 @@
...
+  def card
+    logger.info(">> accessing card #{user.card.ccn}")
+    user.card
   end
 end
 
-def validate_input(ccn)
-  raise ArgumentError unless ccn.match?("[0-9]{4}[\-:\s]?[0-9]{4}[\-:\s]?[0-9]{4}[\-:\s]?[0-9]{4}")
+# User with name, age and card
+class User
```
Comment (denchdev) [inline validate.rb]: Move User class to separate file

File: validate.rb [hunk 4]
```diff
@@ -1,54 +1,170 @@
...
+    else
+      card.validate
+    end
+  end
 end
 
-def normalise(ccn)
-  ccn.gsub(/[\- \s]/, '')
+# Card with number
+class Card
```
Comment (denchdev) [inline validate.rb]: Move class Card to separate file

**Comment by denchdev** [comment]
You should create file for each class and move it to lib folder. In root directory you should create main file and put there all test code.

File: validate.rb [hunk 5]
```diff
@@ -1,54 +1,170 @@
...
-  validate_input(ccn)
-  ccn = normalise(ccn)
-  check_digit = get_check_sum(ccn)
-  (check_digit % TEN).zero?
-rescue ArgumentError
-  puts CC_INPUT_ERROR_MESSAGE
-  false
+# generic interface for User and UserWrapper
+module UserAbstract
+  def initialize(name, age, card) end
```
Comment (denchdev) [inline validate.rb]: Why we should use it?

Comment (AlexanderBugara) [reply validate.rb]: removed

**Comment by denchdev** [comment]
Please use snake case for file names

**Review by denchdev — APPROVED** [review]

**Pull request merged**